FFF-103: Message builder pattern

adlrocha opened the pull request:
Addresses https://github.com/filecoin-project/go-f3/issues/103. Continuation (and depends on) https://github.com/filecoin-project/go-f3/pull/259

This PR:

- Removes the id from gpbft.Participant.
- Changes the signature and moves `MarshalPayloadForSigning` out of `Signature` so ti can be used in `MessageBuilder`. Without this, the serialization performed by the verifier and the message builder didn't match in tests  leading to a nasty bug to find.  
- Uses message builder to allow the signature of message by several hosts.

adlrocha committed:
FFF-103: Split async deliver from PR #188

adlrocha committed:
FFF-103: make networkFor struct private

adlrocha committed:
FFF-103: address minor comments

adlrocha committed:
FFF-103: minor fix

adlrocha committed:
FFF-103: remove redundant local inbox queue

adlrocha committed:
FFF-103: remove ID from gpbft.Participant

adlrocha committed:
FFF-103: minor fix to participant_test

adlrocha committed:
FFF-103: rm network log, and recover BroadcastSynchronous

adlrocha committed:
Merge branch 'fff-103-async-local' into fff-103-id-removal

adlrocha committed:
FFF-103: add message builder pattern for signing

adlrocha committed:
FFF-103: wip fixing tests

adlrocha committed:
Merge branch 'main' into fff-103-async-local

Signed-off-by: Alfonso de la Rocha <adlrocha@tutamail.com>

adlrocha committed:
Merge branch 'fff-103-async-local' into fff-103-id-removal

adlrocha committed:
Merge branch 'fff-103-id-removal' into fff-103-message-builder

adlrocha committed:
FFF-103: remove ID() from host

adlrocha committed:
FFF-103: remove ID() from Reveiver interface

adlrocha committed:
FFF-103: remove networkName from message build

adlrocha committed:
FFF-103: debugging signature generation

adlrocha committed:
FFF-103: adding more logs

adlrocha committed:
FFF-103: address review

adlrocha committed:
Merge branch 'fff-103-async-local' into fff-103-id-removal

adlrocha committed:
Merge branch 'fff-103-id-removal' into fff-103-message-builder

Signed-off-by: Alfonso de la Rocha <adlrocha@tutamail.com>

Kubuxu committed:
Merge branch 'main' into fff-103-async-local

adlrocha committed:
FFF-103: remove logs and getting ready to fix bug

adlrocha committed:
FFF-103: Fix message builder marshaling

adlrocha committed:
FFF-103: minor fix

adlrocha committed:
FFF-103: minor changes

masih commented on gpbft/participant_test.go:
```diff
@@ -86,8 +86,8 @@ func (pt *participantTestSubject) expectBeginInstance() {
 	pt.host.On("NetworkName").Return(pt.networkName).Maybe()
-	pt.host.On("MarshalPayloadForSigning", mock.AnythingOfType("*gpbft.Payload")).
-		Return([]byte(gpbft.DOMAIN_SEPARATION_TAG + ":" + pt.networkName))
+	pt.host.On("MarshalPayloadForSigning", pt.networkName, mock.AnythingOfType("*gpbft.Payload")).
```
I would leave a comment explaining the rationale for `Maybe()`

masih commented on gpbft/vrf.go:
```diff
@@ -14,19 +14,19 @@ type VRFHost interface {
 
-func MakeTicket(beacon []byte, instance uint64, round uint64, source PubKey, host VRFHost) (Ticket, error) {
-	return host.Sign(source, serializeSigInput(beacon, instance, round, host.NetworkName()))
+func MakeTicket(nn NetworkName, beacon []byte, instance uint64, round uint64, source PubKey, host VRFHost) (Ticket, error) {
```
Discussion point for later date: once the number of arguments to a function go above 3 I would start to question what we might be missing in the API/interface design.

Hopefully once the implementation settles we would revisit.

masih reviewed (approved):
This is a PR to an open PR. Once flattened I'll give it another look. For now no blockers from me 👍

adlrocha commented on gpbft/message_builder.go:
```diff
@@ -0,0 +1,114 @@
+
+	PubKey        PubKey
+	PayloadToSign []byte
+	VRFToSign     []byte
```
They definitely shouldn't be exported. I copy-pasted from Kuba's PR and didn't even stop twice to think about it. I will address it in my next commit. Good catch. Thanks!

adlrocha commented on gpbft/message_builder.go:
```diff
@@ -0,0 +1,114 @@
+
+	BeaconForTicket []byte
+
+	Justification *Justification
```
Same as above

adlrocha committed:
FFF-103: use private fields and receiver functions for message builder

Kubuxu committed:
Merge branch 'main' into fff-103-async-local

adlrocha commented on gpbft/message_builder.go:
```diff
@@ -0,0 +1,114 @@
+
+// Build uses the template and a signer interface to build GMessage
+// It is a shortcut for when separated flow is not required
+func (mt MessageBuilder) Build(networkName NetworkName, signer SignerWithMarshaler, id ActorID) (*GMessage, error) {
```
As Kuba mentioned,  he was using this convention to signal unmutable functions, i.e. functions that don't modify the object internally. I kind of like it, so I will keep it like that.

adlrocha commented on gpbft/message_builder.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+// Complete takes the template and signatures and builds GMessage out of them
+func (st SignatureBuilder) Complete(payloadSignature []byte, vrf []byte) *GMessage {
```
I refactor it a bit to make `sign` and `build` private functions of `SignatureBuilder` because right now they are just internally used by `MessageBuilder`

adlrocha commented on gpbft/message_builder.go:
```diff
@@ -0,0 +1,114 @@
+}
+
+// Sign creates signatures for the SigningTemplate, it could live across RPC boundry
+func (st SignatureBuilder) Sign(signer Signer) ([]byte, []byte, error) {
```
Same as above, unless strongly opposed, I am using Kuba's convention of using pointers only for mutable methods.

adlrocha commented on gpbft/message_builder_test.go:
```diff
@@ -0,0 +1,109 @@
+	require.Nil(t, st.VRFToSign)
+}
+
+func TestMessageTemplateWithVRF(t *testing.T) {
```
fixed

adlrocha commented on gpbft/api.go:
```diff
@@ -60,6 +59,13 @@ type Signer interface {
 	Sign(sender PubKey, msg []byte) ([]byte, error)
 }
 
+type MessageMarshaler interface {
```
1. is addressed, but for 2. I thought the same but all my attempts felt a bit dirty. Maybe I am missing some easy way, any proposal? Thanks!

adlrocha committed:
FFF-103: second round of comments

adlrocha commented:
@masih, I think I addressed all of your comments. Let me know if there is any outstanding one. If this is not the case, we can do what we discussed sync of merging this one to the chained PR and have a final review of the whole chain in a single PR. 

Unless @Kubuxu, or any one else wants to review this one before merging to the dependent branch, in that case please let me know before tomorrow. Thanks :pray:

masih commented:
Thanks @adlrocha please feel free to merge this once the test failure is fixed. We can follow up on any outstanding items in the flattened PR 🙏

adlrocha committed:
Merge branch 'fff-103-async-local' into fff-103-message-builder

anorth reviewed (approved):
Thanks. 

Please check that with tracing enabled, the log from a basic simulation test looks similar to before this change. Otherwise LGTM.

I'd prefer you rebase this against main once upstream has merged, rather than combined into one PR, if possible.

adlrocha committed:
FFF-103: fix repeat test to avoid infinite loop

adlrocha committed:
Merge branch 'main' into fff-103-async-local

Signed-off-by: Alfonso de la Rocha <adlrocha@tutamail.com>

adlrocha committed:
Merge branch 'fff-103-async-local' into fff-103-message-builder

Signed-off-by: Alfonso de la Rocha <adlrocha@tutamail.com>

adlrocha committed:
Merge branch 'main' into fff-103-message-builder

Signed-off-by: Alfonso de la Rocha <adlrocha@tutamail.com>

adlrocha committed:
FFF-103: minor fix

adlrocha committed:
Merge branch 'main' into fff-103-message-builder

Signed-off-by: Alfonso de la Rocha <adlrocha@tutamail.com>

adlrocha committed:
FFF-103: adapt moduleClient and PR after rebase

anorth commented:
This did end up breaking the participant ID in trace logs. Fixed in #313.
